src/main/java/htsjdk/samtools/SAMRecordDuplicateComparator.java

```diff
@@ -105,20 +125,20 @@ private static String getLibraryName(final SAMRecord rec) {
             }
         }
 
-        return "Unknown Library";
+        return UNKNOWN_LIBRARY_STRING;
     }
 
     /** Get the library ID for the given SAM record. */
     private short getLibraryId(final SAMRecord rec) {
```

Comment:
I think there are still race conditions here in some (potentially rare) conditions.  

If the records have different sam headers ( and you're not pre-initializing them with the now deprecated constructor)  than the id is dependent on the order that getLibraryID is called in which is a race condition.  Maybe we don't care about that case since we don't expect to have different headers given that you've deprecated that constructor.  

What happened to syncronizing populateTransientAttributes?  

Using a ConcurrentHashMap.calculateIfMissing is a good idea.